# Deep Learning Models

## ** Notebook Content **

1. Introduction


2. Data Preprocessing
    - 2.1 Importing the required packages
    - 2.2 Loading the dataset
    - 2.3 Applgying Feature Scaling
    

3. Artificial Neural Netwoks Model




## 1. Introduction

Explanation...

## 2. Loading data

In [1]:
# Importing the required packages
import sys
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import tensorflow as tf
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LeakyReLU,PReLU,ELU

Using TensorFlow backend.


In [2]:
#check the python version that we are using
print(sys.version)

3.7.7 (default, May  6 2020, 11:45:54) [MSC v.1916 64 bit (AMD64)]


In [3]:
#check the tensorflow version that we are using
tf.__version__

'2.1.0'

### 2.1 Loading the training data

In [4]:
#loading the X_training set
df_train = pd.read_csv('df_train_clean.csv')
df_train.head()

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0          60       RL         65.0     8450   Pave      Reg         Lvl   
1          20       RL         80.0     9600   Pave      Reg         Lvl   
2          60       RL         68.0    11250   Pave      IR1         Lvl   
3          70       RL         60.0     9550   Pave      IR1         Lvl   
4          60       RL         84.0    14260   Pave      IR1         Lvl   

  Utilities LotConfig LandSlope  ... EnclosedPorch 3SsnPorch ScreenPorch  \
0    AllPub    Inside       Gtl  ...             0         0           0   
1    AllPub       FR2       Gtl  ...             0         0           0   
2    AllPub    Inside       Gtl  ...             0         0           0   
3    AllPub    Corner       Gtl  ...           272         0           0   
4    AllPub       FR2       Gtl  ...             0         0           0   

  PoolArea MiscVal  MoSold  YrSold  SaleType  SaleCondition SalePrice  
0        0       0       2    2008        WD         Normal  208500.0  
1        0       0       5    2007        WD         Normal  181500.0  
2        0       0       9    2008        WD         Normal  223500.0  
3        0       0       2    2006        WD        Abnorml  140000.0  
4        0       0      12    2008        WD         Normal  250000.0  

[5 rows x 74 columns]

In [5]:
df_train.shape

(1445, 74)

### 2.2 Loading the test data 

In [6]:
#loading the X_test set
X_test = pd.read_csv('df_test_clean.csv')
X_test.head()

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0          20       RH         80.0    11622   Pave      Reg         Lvl   
1          20       RL         81.0    14267   Pave      IR1         Lvl   
2          60       RL         74.0    13830   Pave      IR1         Lvl   
3          60       RL         78.0     9978   Pave      IR1         Lvl   
4         120       RL         43.0     5005   Pave      IR1         HLS   

  Utilities LotConfig LandSlope  ... OpenPorchSF EnclosedPorch 3SsnPorch  \
0    AllPub    Inside       Gtl  ...           0             0         0   
1    AllPub    Corner       Gtl  ...          36             0         0   
2    AllPub    Inside       Gtl  ...          34             0         0   
3    AllPub    Inside       Gtl  ...          36             0         0   
4    AllPub    Inside       Gtl  ...          82             0         0   

  ScreenPorch PoolArea  MiscVal  MoSold  YrSold  SaleType SaleCondition  
0         120        0        0       6    2010        WD        Normal  
1           0        0    12500       6    2010        WD        Normal  
2           0        0        0       3    2010        WD        Normal  
3           0        0        0       6    2010        WD        Normal  
4         144        0        0       1    2010        WD        Normal  

[5 rows x 73 columns]

In [7]:
X_test.shape

(1459, 73)

## 3. Data Preprocessing

### 3.1 Deleting variables with low correlation

In [8]:
#remove variables with low correlation in the training set
df_train.drop(['MoSold', 'ScreenPorch', '3SsnPorch', 'PoolArea', 'MiscVal', 'YrSold', 'LowQualFinSF', 'MSSubClass',
               'BsmtFinSF2', 'BsmtHalfBath', 'OverallCond'], axis = 1, inplace = True)

df_train.shape

(1445, 63)

In [9]:
#remove variables with low correlation in the test set
X_test.drop(['MoSold', 'ScreenPorch', '3SsnPorch', 'PoolArea', 'MiscVal', 'YrSold', 'LowQualFinSF', 'MSSubClass',
               'BsmtFinSF2', 'BsmtHalfBath', 'OverallCond'], axis = 1, inplace = True)

X_test.shape

(1459, 62)

### 3.2 Treating categorical variables

#### 3.2.1 Convert some categorical variables into dummy

#### a) Training data

In [10]:
#adjustments in the training data
df_train = pd.get_dummies (df_train, columns = ['LotShape', 'LandContour', 'LandSlope', 'BldgType', 'ExterQual', 
                                              'BsmtQual', 'BsmtCond', 'BsmtExposure', 'CentralAir', 'KitchenQual', 
                                              'GarageFinish', 'PavedDrive'])

df_train.shape

(1445, 95)

In [11]:
df_train.columns

Index(['MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Utilities',
       'LotConfig', 'Neighborhood', 'Condition1', 'Condition2', 'HouseStyle',
       'OverallQual', 'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl',
       'Exterior1st', 'Exterior2nd', 'MasVnrArea', 'ExterCond', 'Foundation',
       'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtUnfSF',
       'TotalBsmtSF', 'Heating', 'HeatingQC', 'Electrical', '1stFlrSF',
       '2ndFlrSF', 'GrLivArea', 'BsmtFullBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Functional',
       'Fireplaces', 'GarageType', 'GarageYrBlt', 'GarageCars', 'GarageArea',
       'GarageQual', 'GarageCond', 'WoodDeckSF', 'OpenPorchSF',
       'EnclosedPorch', 'SaleType', 'SaleCondition', 'SalePrice',
       'LotShape_IR1', 'LotShape_IR2', 'LotShape_IR3', 'LotShape_Reg',
       'LandContour_Bnk', 'LandContour_HLS', 'LandContour_Low',
       'LandContour_Lvl', 'LandSlope_Gtl', 'LandSlope_Mod', 'LandSlope_Sev',
 

#### b) Test Data

In [12]:
#adjustments in the test data
X_test = pd.get_dummies (X_test, columns = ['LotShape', 'LandContour', 'LandSlope', 'BldgType', 'ExterQual', 
                                              'BsmtQual', 'BsmtCond', 'BsmtExposure', 'CentralAir', 'KitchenQual', 
                                              'GarageFinish', 'PavedDrive'])

X_test.shape

(1459, 94)

In [13]:
X_test.columns

Index(['MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Utilities',
       'LotConfig', 'Neighborhood', 'Condition1', 'Condition2', 'HouseStyle',
       'OverallQual', 'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl',
       'Exterior1st', 'Exterior2nd', 'MasVnrArea', 'ExterCond', 'Foundation',
       'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtUnfSF',
       'TotalBsmtSF', 'Heating', 'HeatingQC', 'Electrical', '1stFlrSF',
       '2ndFlrSF', 'GrLivArea', 'BsmtFullBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Functional',
       'Fireplaces', 'GarageType', 'GarageYrBlt', 'GarageCars', 'GarageArea',
       'GarageQual', 'GarageCond', 'WoodDeckSF', 'OpenPorchSF',
       'EnclosedPorch', 'SaleType', 'SaleCondition', 'LotShape_IR1',
       'LotShape_IR2', 'LotShape_IR3', 'LotShape_Reg', 'LandContour_Bnk',
       'LandContour_HLS', 'LandContour_Low', 'LandContour_Lvl',
       'LandSlope_Gtl', 'LandSlope_Mod', 'LandSlope_Sev', 'BldgType_1Fam

#### 3.2.2 Convert the ramaining categorical variables into numbers

In [14]:
#adjustments in the training data
from sklearn.preprocessing import LabelEncoder
lencoders = {}

for col in df_train.select_dtypes(include=['object']).columns:
    lencoders[col] = LabelEncoder()
    df_train[col] = lencoders[col].fit_transform(df_train[col])

In [15]:
#check the datatype of df_train to review that all the variables are numbers
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1445 entries, 0 to 1444
Data columns (total 95 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   MSZoning          1445 non-null   int32  
 1   LotFrontage       1445 non-null   float64
 2   LotArea           1445 non-null   int64  
 3   Street            1445 non-null   int32  
 4   Utilities         1445 non-null   int32  
 5   LotConfig         1445 non-null   int32  
 6   Neighborhood      1445 non-null   int32  
 7   Condition1        1445 non-null   int32  
 8   Condition2        1445 non-null   int32  
 9   HouseStyle        1445 non-null   int32  
 10  OverallQual       1445 non-null   int64  
 11  YearBuilt         1445 non-null   int64  
 12  YearRemodAdd      1445 non-null   int64  
 13  RoofStyle         1445 non-null   int32  
 14  RoofMatl          1445 non-null   int32  
 15  Exterior1st       1445 non-null   int32  
 16  Exterior2nd       1445 non-null   int32  


In [16]:
df_train.head()

MSZoning  LotFrontage  LotArea  Street  Utilities  LotConfig  Neighborhood  \
0         3         65.0     8450       1          0          4             5   
1         3         80.0     9600       1          0          2            24   
2         3         68.0    11250       1          0          4             5   
3         3         60.0     9550       1          0          0             6   
4         3         84.0    14260       1          0          2            15   

   Condition1  Condition2  HouseStyle  ...  KitchenQual_Ex  KitchenQual_Fa  \
0           2           2           5  ...               0               0   
1           1           2           2  ...               0               0   
2           2           2           5  ...               0               0   
3           2           2           5  ...               0               0   
4           2           2           5  ...               0               0   

   KitchenQual_Gd  KitchenQual_TA  GarageFinish_Fin  GarageFinish_RFn  \
0               1               0                 0                 1   
1               0               1                 0                 1   
2               1               0                 0                 1   
3               1               0                 0                 0   
4               1               0                 0                 1   

   GarageFinish_Unf  PavedDrive_N  PavedDrive_P  PavedDrive_Y  
0                 0             0             0             1  
1                 0             0             0             1  
2                 0             0             0             1  
3                 1             0             0             1  
4                 0             0             0             1  

[5 rows x 95 columns]

In [17]:
#adjustments in the test data
from sklearn.preprocessing import LabelEncoder
lencoders = {}

for col in X_test.select_dtypes(include=['object']).columns:
    lencoders[col] = LabelEncoder()
    X_test[col] = lencoders[col].fit_transform(X_test[col])

In [18]:
#check the datatype of X_test to review that all the variables are numbers
X_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Data columns (total 94 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   MSZoning          1459 non-null   int32  
 1   LotFrontage       1459 non-null   float64
 2   LotArea           1459 non-null   int64  
 3   Street            1459 non-null   int32  
 4   Utilities         1459 non-null   int32  
 5   LotConfig         1459 non-null   int32  
 6   Neighborhood      1459 non-null   int32  
 7   Condition1        1459 non-null   int32  
 8   Condition2        1459 non-null   int32  
 9   HouseStyle        1459 non-null   int32  
 10  OverallQual       1459 non-null   int64  
 11  YearBuilt         1459 non-null   int64  
 12  YearRemodAdd      1459 non-null   int64  
 13  RoofStyle         1459 non-null   int32  
 14  RoofMatl          1459 non-null   int32  
 15  Exterior1st       1459 non-null   int32  
 16  Exterior2nd       1459 non-null   int32  


In [19]:
X_test.head()

MSZoning  LotFrontage  LotArea  Street  Utilities  LotConfig  Neighborhood  \
0         2         80.0    11622       1          0          4            12   
1         3         81.0    14267       1          0          0            12   
2         3         74.0    13830       1          0          4             8   
3         3         78.0     9978       1          0          4             8   
4         3         43.0     5005       1          0          4            22   

   Condition1  Condition2  HouseStyle  ...  KitchenQual_Ex  KitchenQual_Fa  \
0           1           2           2  ...               0               0   
1           2           2           2  ...               0               0   
2           2           2           4  ...               0               0   
3           2           2           4  ...               0               0   
4           2           2           2  ...               0               0   

   KitchenQual_Gd  KitchenQual_TA  GarageFinish_Fin  GarageFinish_RFn  \
0               0               1                 0                 0   
1               1               0                 0                 0   
2               0               1                 1                 0   
3               1               0                 1                 0   
4               1               0                 0                 1   

   GarageFinish_Unf  PavedDrive_N  PavedDrive_P  PavedDrive_Y  
0                 1             0             0             1  
1                 1             0             0             1  
2                 0             0             0             1  
3                 0             0             0             1  
4                 0             0             0             1  

[5 rows x 94 columns]

### 3.3 Applying Feature Scaling

#### 3.3.1 Training data adjustments

In [20]:
#Normalizing the numerical training data
from sklearn import preprocessing

min_max_scaler = preprocessing.MinMaxScaler()
df_train_scaled = min_max_scaler.fit_transform(df_train)
df_train_scaled = pd.DataFrame(df_train_scaled)

In [21]:
df_train_scaled.head() 

0         1         2    3    4    5         6      7         8   \
0  0.75  0.150685  0.033420  1.0  0.0  1.0  0.208333  0.250  0.285714   
1  0.75  0.202055  0.038795  1.0  0.0  0.5  1.000000  0.125  0.285714   
2  0.75  0.160959  0.046507  1.0  0.0  1.0  0.208333  0.250  0.285714   
3  0.75  0.133562  0.038561  1.0  0.0  0.0  0.250000  0.250  0.285714   
4  0.75  0.215753  0.060576  1.0  0.0  0.5  0.625000  0.250  0.285714   

         9   ...   85   86   87   88   89   90   91   92   93   94  
0  0.714286  ...  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  
1  0.285714  ...  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  
2  0.714286  ...  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  
3  0.714286  ...  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  
4  0.714286  ...  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  

[5 rows x 95 columns]

In [22]:
# Change the columns naming to  retrieve the original titles
df_train_scaled = df_train_scaled.transpose().set_index(df_train.columns).transpose()
#check the changes in the dataset
df_train_scaled.head(10)

MSZoning  LotFrontage   LotArea  Street  Utilities  LotConfig  \
0      0.75     0.150685  0.033420     1.0        0.0        1.0   
1      0.75     0.202055  0.038795     1.0        0.0        0.5   
2      0.75     0.160959  0.046507     1.0        0.0        1.0   
3      0.75     0.133562  0.038561     1.0        0.0        0.0   
4      0.75     0.215753  0.060576     1.0        0.0        0.5   
5      0.75     0.219178  0.059899     1.0        0.0        1.0   
6      0.75     0.184932  0.041057     1.0        0.0        1.0   
7      0.75     0.160959  0.042450     1.0        0.0        0.0   
8      1.00     0.102740  0.022529     1.0        0.0        1.0   
9      0.75     0.099315  0.028605     1.0        0.0        0.0   

   Neighborhood  Condition1  Condition2  HouseStyle  ...  KitchenQual_Ex  \
0      0.208333       0.250    0.285714    0.714286  ...             0.0   
1      1.000000       0.125    0.285714    0.285714  ...             0.0   
2      0.208333       0.250    0.285714    0.714286  ...             0.0   
3      0.250000       0.250    0.285714    0.714286  ...             0.0   
4      0.625000       0.250    0.285714    0.714286  ...             0.0   
5      0.458333       0.250    0.285714    0.000000  ...             0.0   
6      0.875000       0.250    0.285714    0.285714  ...             0.0   
7      0.583333       0.500    0.285714    0.714286  ...             0.0   
8      0.708333       0.000    0.285714    0.000000  ...             0.0   
9      0.125000       0.000    0.000000    0.142857  ...             0.0   

   KitchenQual_Fa  KitchenQual_Gd  KitchenQual_TA  GarageFinish_Fin  \
0             0.0             1.0             0.0               0.0   
1             0.0             0.0             1.0               0.0   
2             0.0             1.0             0.0               0.0   
3             0.0             1.0             0.0               0.0   
4             0.0             1.0             0.0               0.0   
5             0.0             0.0             1.0               0.0   
6             0.0             1.0             0.0               0.0   
7             0.0             0.0             1.0               0.0   
8             0.0             0.0             1.0               0.0   
9             0.0             0.0             1.0               0.0   

   GarageFinish_RFn  GarageFinish_Unf  PavedDrive_N  PavedDrive_P  \
0               1.0               0.0           0.0           0.0   
1               1.0               0.0           0.0           0.0   
2               1.0               0.0           0.0           0.0   
3               0.0               1.0           0.0           0.0   
4               1.0               0.0           0.0           0.0   
5               0.0               1.0           0.0           0.0   
6               1.0               0.0           0.0           0.0   
7               1.0               0.0           0.0           0.0   
8               0.0               1.0           0.0           0.0   
9               1.0               0.0           0.0           0.0   

   PavedDrive_Y  
0           1.0  
1           1.0  
2           1.0  
3           1.0  
4           1.0  
5           1.0  
6           1.0  
7           1.0  
8           1.0  
9           1.0  

[10 rows x 95 columns]

#### 3.3.2 Test data Adjustments

In [23]:
#Normalizing the numerical training data
from sklearn import preprocessing

min_max_scaler = preprocessing.MinMaxScaler()
X_test_scaled = min_max_scaler.fit_transform(X_test)
X_test_scaled = pd.DataFrame(X_test_scaled)

In [24]:
X_test_scaled.head() 

0         1         2    3    4    5         6      7    8         9   \
0  0.50  0.329609  0.184147  1.0  0.0  1.0  0.500000  0.125  0.5  0.333333   
1  0.75  0.335196  0.232124  1.0  0.0  0.0  0.500000  0.250  0.5  0.333333   
2  0.75  0.296089  0.224197  1.0  0.0  1.0  0.333333  0.250  0.5  0.666667   
3  0.75  0.318436  0.154326  1.0  0.0  1.0  0.333333  0.250  0.5  0.666667   
4  0.75  0.122905  0.064121  1.0  0.0  1.0  0.916667  0.250  0.5  0.333333   

   ...   84   85   86   87   88   89   90   91   92   93  
0  ...  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0  
1  ...  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  
2  ...  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  1.0  
3  ...  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  
4  ...  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  

[5 rows x 94 columns]

In [25]:
# Change the columns naming to  retrieve the original titles
X_test_scaled = X_test_scaled.transpose().set_index(X_test.columns).transpose()
#check the changes in the dataset
X_test_scaled.head(10)

MSZoning  LotFrontage   LotArea  Street  Utilities  LotConfig  \
0      0.50     0.329609  0.184147     1.0        0.0        1.0   
1      0.75     0.335196  0.232124     1.0        0.0        0.0   
2      0.75     0.296089  0.224197     1.0        0.0        1.0   
3      0.75     0.318436  0.154326     1.0        0.0        1.0   
4      0.75     0.122905  0.064121     1.0        0.0        1.0   
5      0.75     0.301676  0.154725     1.0        0.0        0.0   
6      0.75     0.262570  0.118085     1.0        0.0        1.0   
7      0.75     0.234637  0.125739     1.0        0.0        1.0   
8      0.75     0.357542  0.157918     1.0        0.0        1.0   
9      0.75     0.273743  0.125703     1.0        0.0        0.0   

   Neighborhood  Condition1  Condition2  HouseStyle  ...  KitchenQual_Ex  \
0      0.500000       0.125         0.5    0.333333  ...             0.0   
1      0.500000       0.250         0.5    0.333333  ...             0.0   
2      0.333333       0.250         0.5    0.666667  ...             0.0   
3      0.333333       0.250         0.5    0.666667  ...             0.0   
4      0.916667       0.250         0.5    0.333333  ...             0.0   
5      0.333333       0.250         0.5    0.666667  ...             0.0   
6      0.333333       0.250         0.5    0.333333  ...             0.0   
7      0.333333       0.250         0.5    0.666667  ...             0.0   
8      0.333333       0.250         0.5    0.333333  ...             0.0   
9      0.500000       0.250         0.5    0.333333  ...             0.0   

   KitchenQual_Fa  KitchenQual_Gd  KitchenQual_TA  GarageFinish_Fin  \
0             0.0             0.0             1.0               0.0   
1             0.0             1.0             0.0               0.0   
2             0.0             0.0             1.0               1.0   
3             0.0             1.0             0.0               1.0   
4             0.0             1.0             0.0               0.0   
5             0.0             0.0             1.0               1.0   
6             0.0             0.0             1.0               1.0   
7             0.0             0.0             1.0               1.0   
8             0.0             1.0             0.0               0.0   
9             0.0             0.0             1.0               1.0   

   GarageFinish_RFn  GarageFinish_Unf  PavedDrive_N  PavedDrive_P  \
0               0.0               1.0           0.0           0.0   
1               0.0               1.0           0.0           0.0   
2               0.0               0.0           0.0           0.0   
3               0.0               0.0           0.0           0.0   
4               1.0               0.0           0.0           0.0   
5               0.0               0.0           0.0           0.0   
6               0.0               0.0           0.0           0.0   
7               0.0               0.0           0.0           0.0   
8               0.0               1.0           0.0           0.0   
9               0.0               0.0           0.0           0.0   

   PavedDrive_Y  
0           1.0  
1           1.0  
2           1.0  
3           1.0  
4           1.0  
5           1.0  
6           1.0  
7           1.0  
8           1.0  
9           1.0  

[10 rows x 94 columns]

### 3.4 Getting the dependent and independent variables

In [26]:
df_train_scaled.shape

(1445, 95)

In [27]:
#Getting the Dependent and Independent variables
X_train = df_train_scaled.iloc[:, :-1] #all lines, all columns except the last one
y_train = df_train_scaled.iloc[:, 94] #all lines, only the last column

In [28]:
#check the shaape of X_train and y_train
X_train.shape, y_train.shape

((1445, 94), (1445,))

Explanation...

## 4. Building a predictive model applying Artificial Neural Networks (ANN) 

### 4.1 Building the ANN

In [29]:
## Initializing the ANN
#create the object for the ANN
ann = tf.keras.models.Sequential()

In [30]:
## Adding the input layer and the first hidden layer
#add a simple connected layer
ann.add(tf.keras.layers.Dense(units=50, activation='relu'))

In [31]:
## Adding the second hidden layer
#add a simple connected layer
ann.add(tf.keras.layers.Dense(units=50, activation='relu'))

In [32]:
## Adding the third hidden layer
#add a simple connected layer
ann.add(tf.keras.layers.Dense(units=50, activation='relu'))

In [33]:
## Adding the output layer
#add the output layer
ann.add(tf.keras.layers.Dense(units=1))

### 4.2 Training the ANN model

In [34]:
#compiling the ANN
ann.compile(optimizer = 'adam', loss = 'mean_squared_error')

In [35]:
#training the ANN
ann.fit(X_train.values, y_train.values, batch_size = 32, epochs = 100, validation_split=0.2)

Train on 1156 samples, validate on 289 samples
Epoch 1/100
1156/1156 [==============================] - 1s 563us/sample - loss: 0.1374 - val_loss: 0.0947
Epoch 2/100
1156/1156 [==============================] - 0s 69us/sample - loss: 0.0592 - val_loss: 0.0726
Epoch 3/100
1156/1156 [==============================] - 0s 72us/sample - loss: 0.0466 - val_loss: 0.0600
Epoch 4/100
1156/1156 [==============================] - 0s 76us/sample - loss: 0.0331 - val_loss: 0.0357
Epoch 5/100
1156/1156 [==============================] - 0s 71us/sample - loss: 0.0198 - val_loss: 0.0194
Epoch 6/100
1156/1156 [==============================] - 0s 69us/sample - loss: 0.0116 - val_loss: 0.0127
Epoch 7/100
1156/1156 [==============================] - 0s 67us/sample - loss: 0.0073 - val_loss: 0.0082
Epoch 8/100
1156/1156 [==============================] - 0s 64us/sample - loss: 0.0039 - val_loss: 0.0062
Epoch 9/100
1156/1156 [==============================] - 0s 69us/sample - loss: 0.0028 - val_loss: 0.005

1156/1156 [==============================] - 0s 79us/sample - loss: 9.8279e-05 - val_loss: 0.0010
Epoch 75/100
1156/1156 [==============================] - 0s 74us/sample - loss: 9.1311e-05 - val_loss: 0.0011
Epoch 76/100
1156/1156 [==============================] - 0s 69us/sample - loss: 1.0047e-04 - val_loss: 8.6951e-04
Epoch 77/100
1156/1156 [==============================] - 0s 62us/sample - loss: 5.5192e-05 - val_loss: 9.7343e-04
Epoch 78/100
1156/1156 [==============================] - 0s 71us/sample - loss: 8.1390e-05 - val_loss: 0.0010
Epoch 79/100
1156/1156 [==============================] - 0s 67us/sample - loss: 9.5482e-05 - val_loss: 9.0019e-04
Epoch 80/100
1156/1156 [==============================] - 0s 69us/sample - loss: 7.0895e-05 - val_loss: 9.0753e-04
Epoch 81/100
1156/1156 [==============================] - 0s 69us/sample - loss: 7.7315e-05 - val_loss: 9.1676e-04
Epoch 82/100
1156/1156 [==============================] - 0s 69us/sample - loss: 6.5049e-05 - val_loss: 8

### 4.3 Predicting the results

#### 4.3.1 Predict the results of the training set 

In [36]:
#predict the results of the train set
y_pred_1 = ann.predict(X_train)
np.set_printoptions(precision=2)
print(y_pred_1)

[[1.01]
 [1.  ]
 [1.01]
 ...
 [0.99]
 [1.01]
 [1.  ]]


In [37]:
#converting the results into a dataframe
y_pred_1_df = pd.DataFrame(y_pred_1)

#export the results to csv
y_pred_1_df.to_csv('y_pred_ann_train.csv', index=False)

#### 4.3.2 Predict the results of the test set 

In [38]:
#predict the test results of the test set
y_pred_2 = ann.predict(X_test_scaled)
np.set_printoptions(precision=2)
print(y_pred_2)

[[ 0.33]
 [ 0.22]
 [ 0.2 ]
 ...
 [ 0.12]
 [ 0.23]
 [-0.01]]


In [49]:
#revert MinMaxScaler to get real data
min_max_scaler.inverse_transform(y_pred_2)

ValueError: non-broadcastable output operand with shape (1459,1) doesn't match the broadcast shape (1459,94)

In [48]:
#converting the results into a dataframe
#y_pred_2_df = pd.DataFrame(y_pred_2)

#export the results to csv
#y_pred_2_df.to_csv('y_pred_ann_test.csv', index=False) 

### 4.4 Check the performance of the model

#### Evaluating the predicted training results

In [40]:
import sklearn.metrics as metrics
#Compute R-squared for the y_train and y_pred_2 (training results)
print("R-squared (Training Results) =", metrics.r2_score(y_train, y_pred_1))
#Compute Mean-Squared-Error (MSE) for the y_train and y_pred_1 (training results)
print("MSE (Training Results) =", metrics.mean_squared_error(y_train, y_pred_1))

R-squared (Training Results) = 0.9973680852825622
MSE (Training Results) = 0.00020041639351665462


In [41]:
from sklearn.metrics import mean_squared_error
from math import sqrt

#Compute Root-Mean-Squared-Error (RMSE) for the y_train and y_pred_1 (training results)
print("RMSE (Training Results)=", sqrt(mean_squared_error(y_train, y_pred_1)))

RMSE (Training Results)= 0.01415684970311738


### 4.5 Create a submission in the Kaggle competition

The last step in our analysis is to merge the "Id" column of the test set with the predicted results (y_pred_4) in order to make a submission in the Kaggle competition. One of the Kaggle's rules is to upload a file with 1459 prediction rows and a header row with the "Id" and "SalePrice" columns.

So, we are going to compile this information in one dataframe and create a submission in the Kaggle Leaderboard.

Let's do it!

In [42]:
#upload the original test data (we need to recover the 'Id' column)
df_test_original = pd.read_csv('test.csv')
df_test_original.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  

[5 rows x 80 columns]

In [43]:
#Creating the "Id_test" dataframe to merge the Id column of the test set with the y_pred_4 results later 
Id_test = df_test_original[['Id']]

In [44]:
#Merging the "Id_test" set with the y_pred_4_df results 
#ann_final_results = pd.concat([Id_test, y_pred_2_df], axis=1,)
#ann_final_results.head()

In [45]:
#check the shape of the ann_training_final_results dataframe
#ann_final_results.shape

In [46]:
#rename the columns of the  rf_final_results dataframe
#ann_final_results.columns = ['Id', 'SalePrice']
#ann_final_results.columns

In [47]:
#export the rf_final_results and y_pred_4 to csv
#ann_final_results.to_csv('ann_final_results.csv', index=False)  